### ISSUE TITLE
🌀Insoluble

### ISSUE THREAD

**Issue by kranzky** [NaNoGenMo/2018#28]
OK, I'm going to _try_ this year.

I want to start with the NaNoWriMo novel I wrote in 2015, which is [here](https://github.com/kranzky/insoluble/blob/master/insoluble.md). I'll tag it by hand, then infer some kind of stochastic model to capture its structure, then use that, with a bunch of models trained on Gutenberg corpora, to generate something new. I want to do it all in Rust, which I'll be learning along the way.

 I have a terrible track record of completing hobby projects.

**Comment by kranzky** [comment]
https://github.com/kranzky/insoluble

**Comment by kranzky** [comment]
OK, first decision: I'll use Ruby and C. Not Rust. Why burden myself?

**Comment by kranzky** [comment]
Baby steps. Here's kinda-sorta what I plan to do:

- Build a model that associates words in three consecutive sentences, and use this to generate a list of keywords that should be in the middle sentence, given two adjacent sentences. Train that on the [Gutenberg Corpus](http://web.eecs.umich.edu/~lahiri/gutenberg_dataset.html).
- Build a model that can generate a sentence given a bunch of keywords that should appear in that sentence. Train that on the same corpus.
- Use my novel from NaNoWriMo 2015 to generate a new novel, by taking sentence pairs from my novel and giving them to the first model above to yield a list of keywords, then generating hundreds of candidate sentences from the keywords using the second model above, then compare the generations with the actual sentence from my novel, and use some heuristic to select the "best" generated sentence.

I'll do all this in Ruby, using [Sooth](https://github.com/kranzky/sooth). Later, if time permits, I would like to write a new crazy grammatical inference engine that I've been thinking about. We'll see...

**Comment by kranzky** [comment]
Wrote some code to process the books in the Gutenberg Corpus. It finds chapters, then breaks the paragraphs in each chapter down into individual sentences, then tags those as dialogue or exposition. After running for about 30 minutes it yields 1947424 paragraphs from 33010 chapters taken from 1653 books. This will be my training data.

I also processed Insoluble to yield a template, which I will use to constrain generation. Here's a snippet:

```
CHAPTER
exposition:I awoke.
exposition:It was far too dark.
exposition:It occurred to me that I'd forgotten to set the alarm, and Julie had neglected to wake me from oversleeping.
exposition:The television was blaring.
exposition:One of the shopping channels judging by the cadence of the presenter's voice.
exposition:I wandered from the bedroom to the kitchen.
exposition:A mess of pots and pans, a half-empty glass of wine, the smell of cooking.
PARAGRAPH
dialogue:"Julie?"
exposition:I called in a hoarse whisper.
dialogue:"Are you home?"
```

The next step will be to process the template, using a model trained on the Gutenberg corpus, to replace each sentence with a list of keywords (which will then be used to constrain generation of new sentences).

**Comment by kranzky** [comment]
After much wailing and gnashing of teeth, I managed to use the Gutenberg Corpus to generate keywords from the Insoluble template. Here are the results, for the snippet of the template shown above:

```
CHAPTER
exposition;2:CONSTRUCT CONNECTION LACKED HUNTED CLAIM
exposition;5:CHASING SHUTTING NAILS RUBS SOIL
exposition;17:OCCURRED NEGLECTED AMISS INDEED EXTRACTED
exposition;4:HOMUNCULUS ETERNITY SECURELY DOZING SLIPPERY
exposition;10:CADENCE JUDGING SLIDE SELECTED DISSUADE
exposition;7:WANDERED DEEPEN BEDROOM CONSOLED BULGING
exposition;12:POTS WINE COOKING WHEREVER SMELL
PARAGRAPH
dialogue;1:HOLY BLOOD NURSING SACRIFICING DEFENDING
exposition;6:HOARSE AGREED PASSION CRIES DEAREST
dialogue;3:ACKNOWLEDGEMENT ACCOUNT ADJOINING EXCURSIONS SHOCKED
```

The next step will be to build language models from the Gutenberg Corpus to allow this template to be fleshed out; essentially I will need to be able to generate a sentence that is guaranteed to contain at least one, but hopefully all three, of the keywords.

**Comment by kranzky** [comment]
I built the language model for generating sentences that are constrained by a list of keywords. I call it a fractal language model. As a quick test, I trained it using just these four keywords: TELEPHONE, BENT, SCREAMED, LIPS. The training data was 938 sentences from the Gutenberg Corpus (I selected 100 random books, and those sentences contained at least one of the four keywords). I then asked it to generate sentences containing 1, 2 or 3 keywords, with these hand-picked results:

```
THERE _SCREAMED_ AND DO OURSELVES PERCHANCE YOU ARE DISTORTED BY _TELEPHONE_ FROM SCOTLAND YARD MY BUSINESS IS TO SPEAK SLOWLY IF HE EXPLAINED TO HIS _LIPS_

BY _TELEPHONE_ CORCORAN _SCREAMED_ AND CLUNG TO CHRISTINE WHO WERE HOLDING HER

MR TUPMAN'S _LIPS_ AS IF SHE _BENT_ AND HIS OWN SAKE AND CUTTING AWAY THE ARM CARESSINGLY

THE _LIPS_ WERE DARKENING FROM LEYDEN WITH AMUSEMENT AT LAST THE GIRLS _SCREAMED_ TO THE BRINK

BEFORE THE MOUTH WITH ME _SCREAMED_ WITH HORROR WHILE EVERY TUG OF THEIR RAVENOUS HATE AND LOOKING ROUND A WOLF UPON A TURN OF AN INSTANT LATER AND RAGE THEY LEAPT UP THE SHOUTS OF THE _TELEPHONE_
```

This is good enough for now.

**Comment by kranzky** [comment]
Here's an example of using the fractal language model to flesh out the keywords shown [above](https://github.com/NaNoGenMo/2018/issues/28#issuecomment-436523073), trained on a mere 15051 sentences from the Gutenberg Corpus:

```
CHAPTER
exposition:SCARCE HAD SLEPT IN DEEP SLUMBER
exposition:IMMEDIATELY CONCERNED BUT HE AWOKE THE SLEEPING MAN
exposition:OUTSIDE THROUGH A DOZEN COOKIES CRUMBLED AWAY FROM HER WITH THAT FOLLOWED THEODORA FELL ASLEEP
exposition:PHYLLIS WAKE OF A LITTLE AHEAD
exposition:THE HOARSE A GRUFF WITH THEM YESTERDAY'S WAR OF WORDS
exposition:IN HER AMPLE TIME FLUENT FRENCH DISHES
exposition:AND TANGLED THE DIRECTORS OF THE WINE WITHOUT RESIN AND THE GLASS AND AGAIN AND PANS OF THE FORCES
PARAGRAPH
dialogue:TARZAN
exposition:AGAIN EXULTING WHISPER A HOARSE VOICE ROSE
dialogue:REMARKABLY SENSITIVE CHARACTERISTICS
```

The next step is to train on all of the corpus...

**Comment by kranzky** [comment]
Still working on this! Not much more to report; just fine-tuning generated text to hit the 50K minimum and to implement a minimal blacklist (I tested by blacklisting the word "THE" and the results were quite interesting, so I might do several generations with different common words omitted).

Here's a quick example of what I've got so far. The original text, from my 2016 NaNoWriMo effort:

```
The room beyond is dark. He peers inside, his eyes slowly adjusting to the gloom. He fancies that he sees a shadowy figure towering before him. The shape twists slightly as a breeze blows in from outside, the beams of the ceiling creaking in protest. A body is suspended in mid-air, arms spread apart, palms facing outward. Beckoning him to enter.

"Jesus!"

Mike waves his hand urgently to activate the artificial lighting. The room quickly brightens in response.

"Christopher!"

Mike rushes into the room, his colleague hanging above him, long dead. Julie falls to her knees in the vestibule, retching and whimpering, eyes tightly shut. Arthur explodes into life from his hiding hole and attempts to enter the room.

"Stop him," cries Mike to Julie. "He'll call the police!"

Julie, now curled up into a softly sobbing ball, doesn't respond. Mike scans the room quickly, looking for something, anything to use to dissuade Arthur from his instinctive response to a dead human body. He rushes into the messy kitchenette, almost overcome by its stink, and continues into the bedroom beyond. A cricket bat leans against the far wall. Consumed with murderous intent, Mike vaults over the bed in one smooth motion to seize it.
```

And text generated by models that never saw the original:

```
AND CONFINES THE HEADLIGHTS IN THE DARK. WHEN THEY WERE BLOODSHOT HE SWAYED TO SPARKLE OF THE HEAVY LIDS. HE SHOOK HIS GREAT MOUNTAIN TALL AND PEERS AND GRACEFUL. IT WAS BLOWING VERY HARD AT THAT THE WALLS OF RED CLOTH AND SOUNDED BUT THE FLOOR. HE SAID SIMPLY FOLDED HER FACE SOME INWARD LAUGHTER AND FOLDING WINDOWS SHE EXPLAINED. WHEN SUSPENDED TO FOLLOW HIS EXAMPLE.

GOD.

THE SMOULDERING PIPE WITH FOAM SEEMED ALMOST PRESSED LEVEL BY THE CIGAR. MIKE RELIED UPON A BLACK WAVES.

WREN.

THE SPIKES TOWERED BEHIND THE TOWERING THEY WERE THICKLY STREWN WITH RUSHES. HER ELBOWS MIKE THIS IS THE ROOM ASKED IN A LITTLE BLOODSHOT. THE UNIVERSITY AND FALLS JULIE IT WAS NEEDFUL THAT HOUSE DUG.

CLEAR SHRILL CRY I GO HOME AGAIN SHRIEKS. HE'S GONE SAID MIKE

HIS LIP HAD A MAN LAUGHED AGAIN IN HIS HAIR. HE SHOOK WITH CURIOUS BEINGS WHO CAME FROM THE CIRCLE OF HIS BALL JULIE IS MUCH LESS DISPLEASING SEX. SHE ASKED ARTHUR MADE MIKE COULD NOT IN THE BEST IF ONLY REMEMBER ARTHUR'S FACE. I COULD SEE THAT CONCERNED ABOUT HIS FRECKLED. THAT RAISED HER THE WALL IS VOUCHSAFED FEW MINUTES AT CRICKET IN THE INNER SURFACE OF THE TIME.
```

**Comment by kranzky** [comment]
I'm happy with my [blacklist](https://raw.githubusercontent.com/kranzky/insoluble/master/blacklist.txt), but not so happy with the keywords, so I'll spend a few days going around in circles; generating new keywords, using them to generate the novel, updating the blacklist based on the results, rinse-and-repeat.

**Comment by kranzky** [comment]
Not so fast @hugovk... although the generation is done I still have some work to do to tidy up the results. I'm planning on inferring two models to do this: the first will fix punctuation by predicting the separator that should come between two words, while the second will fix word case by predicting the correct case for a word, given the preceding word and punctuation. The end result should look a bit more like a novel instead of LINES OF SHOUTING.

**Comment by hugovk** [comment]
OK, label removed! Ping me when you're ready :)

**Comment by kranzky** [comment]
Had an idea for improving keyword generation, so will do that and see if the results are better.

**Comment by kranzky** [comment]
Finished that work; to recap, here's the first few lines from the start of the second chapter in the source template:

```
CHAPTER
exposition:I awoke.
exposition:It was far too dark.
exposition:It occurred to me that I'd forgotten to set the alarm, and Julie had neglected to wake me from oversleeping.
exposition:The television was blaring.
exposition:One of the shopping channels judging by the cadence of the presenter's voice.
exposition:I wandered from the bedroom to the kitchen.
exposition:A mess of pots and pans, a half-empty glass of wine, the smell of cooking.
PARAGRAPH
dialogue:"Julie?"
exposition:I called in a hoarse whisper.
dialogue:"Are you home?"
```

Here are the keywords we now extract:

```
CHAPTER
exposition;2:AWOKE LAYERS STRUCTURES
exposition;5:DARK OPAQUE ACT
exposition;17:OCCURRED FORGOTTEN JULIE
exposition;4:JE POWERFUL DREAMS
exposition;10:CADENCE VOICE SHOPPING
exposition;7:WANDERED INTERVIEWED DISCUSSING
exposition;12:AGGRESSIVE EXECUTED POTS
PARAGRAPH
dialogue;1:JULIE PROFESSIONAL MENIAL
exposition;6:WHISPER HOARSE RUSHES
dialogue;3:HOME PRESENTATION SOD
```

And here is a generation using those keywords:

```
CHAPTER
exposition;1,1:I AWOKE ME
exposition;2,2:THIS DARK THAT ACT CAME A BRIDE'S-MAID
exposition;3,1:THE INTERRUPTION OCCURRED AS NOTICEABLE AS GENTLE JULIE I HAVE HAD ESCAPED THE MOMENT THAT OUR FATIGUES FORGOTTEN
exposition;1,1:BUT HE PRESSED BY DREAMS
exposition;3,2:RECKLESSLY SHOPPING TO THIS WAS LIKE A PLAINTIVE CADENCE AND DELICATE VOICE
exposition;2,1:FIRST TIME DISCUSSING I WANDERED FROM HER HEART
exposition;3,0:SHE HAD MET WITH POTS TO HEAR THAT THE AGGRESSIVE AFTERWARDS EXECUTED
PARAGRAPH
dialogue;1,2:MY PROFESSIONAL REGRET
exposition;2,0:THE MAN OF RUSHES WE WHISPER
dialogue;1,0:YOU SAFE HOME
```

I think the results are more interesting, yet still relevant.

Onward!

**Comment by kranzky** [comment]
I've finished the repair script, and am currently working on an improvement to the fractal language model, which is unfortunately taking _days_ to run on an AWS high memory instance. So I may need to ditch that.

**Comment by kranzky** [comment]
Working on multithreading, as my language model is taking too long to train...

**Comment by kranzky** [comment]
@hugovk please mark this one as completed now :)

The generated novel, along with an explanation of how it all works, is here: https://github.com/kranzky/insoluble/blob/master/README.md

**Comment by hugovk** [comment]
Labelled, and congratulations!
